### DISCUSSION TITLE
How to use csv-async crate with Rocket?

### DISCUSSION THREAD

**Discussion by padraigconnolly** [rwf2/Rocket#2601] [Questions] [answered]
I'm currently working on a small project where I open a CSV file as part of a GET request with Rocket. Of course I am trying to do this asynchronously so I'm trying to use [csv-async](https://crates.io/crates/csv-async) and the following is my effort so far:
```rust
#[macro_use] extern crate rocket;

use rocket_dyn_templates::{Template, context};

#[get("/")]
async fn hello() -> Template {
    let mut rdr: csv_async::AsyncReader<Result<async_std::fs::File, std::io::Error>> = csv_async::AsyncReader::from_reader(async_std::fs::File::open("csv/entities.csv").await);
    Template::render("index", context! {
        name: "Derek",
        items: vec!["One", "Two", "Three"],
    })
}

#[launch]
fn rocket() -> _ {
    rocket::build()
    .mount("/", routes![hello])
    .attach(Template::fairing())
}
```
Now the line ` let mut rdr: csv_async::AsyncReader<Result<async_std::fs::File, std::io::Error>> = csv_async::AsyncReader::from_reader(async_std::fs::File::open("csv/entities.csv").await);` is causing me a lot of headache where the following issue prints:
```
   Compiling hardware-tracker v0.1.0 (/home/pjconnol/rust/hardware-tracker)
error[E0277]: the trait bound `Result<async_std::fs::File, std::io::Error>: rocket::futures::AsyncRead` is not satisfied
   --> src/main.rs:7:124
    |
7   |     let mut rdr: csv_async::AsyncReader<Result<async_std::fs::File, std::io::Error>> = csv_async::AsyncReader::from_reader(async_std::fs::File::open("csv/entities.csv").await);
    |                                                                                        ----------------------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `rocket::futures::AsyncRead` is not implemented for `Result<async_std::fs::File, std::io::Error>`
    |                                                                                        |
    |                                                                                        required by a bound introduced by this call
    |
    = help: the following other types implement trait `rocket::futures::AsyncRead`:
              &[u8]
              &async_io::Async<T>
              &async_std::fs::File
              &async_std::net::TcpStream
              &async_std::os::unix::net::UnixStream
              &mut T
              AllowStdIo<T>
              Box<T>
            and 32 others
note: required by a bound in `AsyncReader::<R>::from_reader`
   --> /home/pjconnol/.cargo/registry/src/index.crates.io-6f17d22bba15001f/csv-async-1.2.6/src/async_readers/ardr_futures.rs:123:8
    |
123 |     R: io::AsyncRead + Unpin + Send + 'r,
    |        ^^^^^^^^^^^^^ required by this bound in `AsyncReader::<R>::from_reader`
...
159 |     pub fn from_reader(rdr: R) -> AsyncReader<R> {
    |            ----------- required by a bound in this associated function

For more information about this error, try `rustc --explain E0277`.
error: could not compile `hardware-tracker` (bin "hardware-tracker") due to previous error
```

So my question is, am I using csv-async and Rocket incorrectly together or is this a bug where these two crates are incompatible with each other because of `AsyncRead`?

**Comment by robertwayne** [comment, marked as answer]
I'm not familiar with `csv-async`, but from a glance, what is your intent when passing `AsyncReader` a `Result<T, E>`?

The problem you are encountering is that `Result<T, E>` doesn't impl `AsyncRead`. You cannot impl this either, as `Result` is defined in Rust core. Of course, you could use a NewType  to get around it, but all of that can be sidestepped by just structuring your code differently.

```rust
#[get("/")]
async fn hello() -> Template {
    // handle for errors first
    let Ok(file) = tokio::fs::File::open("csv/entities.csv").await else {
        todo!("return 404 template or whatever")
    };

    // now we can use the file however we like
    let mut reader = csv_async::AsyncReader::from_reader(file);

    Template::render(
        "index",
        context! {
            name: "Derek",
            items: vec!["One", "Two", "Three"],
        },
    )
}
```

Also, one thing I notice is you are using `async_std::File`, etc. Rocket is based off of Tokio, and runs a Tokio runtime _(as well as exports Tokio)_. You should be using the `tokio` feature for `csv-async` and using `tokio::fs::File` instead. You can see I made those changes in the code above already.

**Reply by padraigconnolly** [reply]
Ah ok I think I misread a csv-async example I was trying to implement alongside rocket, using Tokio makes a lot more sense, thanks for your help, your example is a lot cleaner

**Reply by padraigconnolly** [reply]
Hmm, it looks like csv-async doesn't work well with tokio (I can't even get their sample code to work sadly). I think for the moment, I'll just access the file using tokio i/o and then parse the csv myself. Thanks for the help!
